File: lib/tools/instagram.ts
import { env } from "../config";

const GRAPH_API = "https://graph.facebook.com/v21.0";

export class InstagramError extends Error {}

export function instagramConfigured(): boolean {
  return Boolean(env.IG_ACCESS_TOKEN && env.IG_BUSINESS_ACCOUNT_ID);
}

export async function publishImage(
  imageUrl: string,
  captionWithHashtags: string,
): Promise<{ mediaId: string; permalink: string }> {
  if (!instagramConfigured()) {
    throw new InstagramError("Instagram credentials not configured");
  }

  // 1. Create container
  const createResp = await fetch(`${GRAPH_API}/${env.IG_BUSINESS_ACCOUNT_ID}/media`, {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      image_url: imageUrl,
      caption: captionWithHashtags,
      access_token: env.IG_ACCESS_TOKEN,
    }),
  });
  if (!createResp.ok) throw new InstagramError(`create failed: ${await createResp.text()}`);
  const { id: containerId } = (await createResp.json()) as { id: string };

  // 2. Poll status
  for (let i = 0; i < 15; i++) {
    const sResp = await fetch(
      `${GRAPH_API}/${containerId}?fields=status_code&access_token=${env.IG_ACCESS_TOKEN}`,
    );
    const sBody = (await sResp.json()) as { status_code?: string };
    if (sBody.status_code === "FINISHED") break;
    if (sBody.status_code === "ERROR") {
      throw new InstagramError(`container processing error: ${JSON.stringify(sBody)}`);
    }
    await new Promise((r) => setTimeout(r, 2000));
  }

  // 3. Publish
  const pubResp = await fetch(`${GRAPH_API}/${env.IG_BUSINESS_ACCOUNT_ID}/media_publish`, {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      creation_id: containerId,
      access_token: env.IG_ACCESS_TOKEN,
    }),
  });
  if (!pubResp.ok) throw new InstagramError(`publish failed: ${await pubResp.text()}`);
  const { id: mediaId } = (await pubResp.json()) as { id: string };

  // 4. Permalink
  const linkResp = await fetch(
    `${GRAPH_API}/${mediaId}?fields=permalink&access_token=${env.IG_ACCESS_TOKEN}`,
  );
  const linkBody = (await linkResp.json()) as { permalink?: string };
  return { mediaId, permalink: linkBody.permalink ?? "" };
}

File: src/tools/instagram.py
import asyncio
import logging

import httpx

from ..config import settings

logger = logging.getLogger(__name__)

GRAPH_API = "https://graph.facebook.com/v21.0"


class InstagramError(Exception):
    pass


async def publish_image(image_url: str, caption_with_tags: str) -> dict:
    """Publish a single-image post to Instagram.

    Returns dict with `media_id` and `permalink`.
    Raises InstagramError on failure.
    """
    if not settings.ig_access_token or not settings.ig_business_account_id:
        raise InstagramError("Instagram credentials not configured")

    async with httpx.AsyncClient(timeout=60.0) as client:
        # Step 1: Create media container
        create_resp = await client.post(
            f"{GRAPH_API}/{settings.ig_business_account_id}/media",
            data={
                "image_url": image_url,
                "caption": caption_with_tags,
                "access_token": settings.ig_access_token,
            },
        )
        if create_resp.status_code != 200:
            raise InstagramError(f"create container failed: {create_resp.text}")
        container_id = create_resp.json()["id"]
        logger.info("IG container created: %s", container_id)

        # Step 2: Poll container status until ready (max ~30s)
        for _ in range(15):
            status_resp = await client.get(
                f"{GRAPH_API}/{container_id}",
                params={
                    "fields": "status_code",
                    "access_token": settings.ig_access_token,
                },
            )
            status = status_resp.json().get("status_code")
            if status == "FINISHED":
                break
            if status == "ERROR":
                raise InstagramError(f"container processing error: {status_resp.text}")
            await asyncio.sleep(2)
        else:
            raise InstagramError("container never reached FINISHED status")

        # Step 3: Publish
        publish_resp = await client.post(
            f"{GRAPH_API}/{settings.ig_business_account_id}/media_publish",
            data={
                "creation_id": container_id,
                "access_token": settings.ig_access_token,
            },
        )
        if publish_resp.status_code != 200:
            raise InstagramError(f"publish failed: {publish_resp.text}")
        media_id = publish_resp.json()["id"]

        # Step 4: Get permalink
        perma_resp = await client.get(
            f"{GRAPH_API}/{media_id}",
            params={
                "fields": "permalink",
                "access_token": settings.ig_access_token,
            },
        )
        permalink = perma_resp.json().get("permalink", "")

        logger.info("Published IG media %s: %s", media_id, permalink)
        return {"media_id": media_id, "permalink": permalink}
